### PR TITLE
Feature/bodies and receipts request sizer

### PR DESCRIPTION
**Pull request by asdacap**
- Besu and reth does not seems to have soft limits on their bodies and receipts message, meaning we may receipt messages up to 20MB in size. For reference, geth and nethermind send up to 3MB only. 
- This is not great as larger message are harder to pool by dotnetty, which may allocate more memory to create more arena. (default arena size is 32MB with the default 2GB memory budget).  
- Also, they tend to timeout as they are trying to send very large message.
- This PR add a per-peer request sizer, which adjust the requests size of get block body and get receipts so that it will reduce the request size if previous request take too long or is too big.
- This reduces 99p size of block body and receipt from about 20MB to 8MB, which seems to eliminate very large memory allocation when connecting only to besu.

Screenshot_2023-08-10_18-20-59

## Changes

- Add a request sizer which changes body and receipt request size. 

#### What types of changes does your code introduce?

- [X] Bugfix (a non-breaking change that fixes an issue)
- [X] Optimization

#### Requires testing

- [X] Yes
- [ ] No

#### If yes, did you write tests?

- [X] Yes
- [ ] No

#### Requires explanation in Release Notes

- [X] Maybe?

Slightly reduce memory allocation during old bodies and receipts.

### CODE REVIEW

File: src/Nethermind/Nethermind.Core/RequestSizer/LatencyAndMessageSizeBasedRequestSizer.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    /// <summary>
+    /// Adjust the request size depending on the latency and response size. Accept a list as request which will be capped.
+    /// If the response size is too large, reduce request size.
+    /// If the latency is above watermark, reduce request size.
+    /// If the latency is below watermark and response size is not too large, increase request size.
+    /// </summary>
+    /// <param name="request"></param>
+    /// <param name="func"></param>
+    /// <typeparam name="T">response type</typeparam>
+    /// <typeparam name="TR">request type</typeparam>
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/RequestSizer/LatencyAndMessageSizeBasedRequestSizer.cs:53]: TRequest, TResponse please

File: src/Nethermind/Nethermind.Core/RequestSizer/LatencyAndMessageSizeBasedRequestSizer.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    /// <returns></returns>
+    public async Task<T> Run<T, TR>(IReadOnlyList<TR> request, Func<IReadOnlyList<TR>, Task<(T, long)>> func)
+    {
+        return await _requestSizer.Run(async (adjustedRequestSize) =>
+        {
+            Stopwatch sw = Stopwatch.StartNew();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/RequestSizer/LatencyAndMessageSizeBasedRequestSizer.cs:59]: maybe new Stopwatch API:  https://www.youtube.com/watch?v=NTz99yN2urc

File: src/Nethermind/Nethermind.Core/RequestSizer/LatencyAndMessageSizeBasedRequestSizer.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            }
+
+            if (duration > _upperLatencyWatermark)
+            {
+                return (result, AdaptiveRequestSizer.Direction.Decrease);
+            }
+
+            if (
+                request.Count >= adjustedRequestSize // If the original request size is low, increasing wont do anything
+                && duration < _lowerLatencyWatermark)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/RequestSizer/LatencyAndMessageSizeBasedRequestSizer.cs:74]: maybe some threshold where it doesn't make sense to increase if we are near it?

File: src/Nethermind/Nethermind.Core/RequestSizer/LatencyBasedRequestSizer.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    /// <typeparam name="T"></typeparam>
+    /// <returns></returns>
+    public async Task<T> MeasureLatency<T>(Func<int, Task<T>> func)
+    {
+        return await _requestSizer.Run(async (requestSize) =>
+        {
+            Stopwatch sw = Stopwatch.StartNew();
+            T result = await func(requestSize);
+            TimeSpan duration = sw.Elapsed;
+            if (duration < _lowerWatermark)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/RequestSizer/LatencyBasedRequestSizer.cs:43]: maybe some threshold where it doesn't make sense to increase if we are near it?

File: src/Nethermind/Nethermind.Network/P2P/ProtocolHandlers/SyncPeerProtocolHandlerBase.cs [hunk 5]
```diff
@@ -51,7 +54,25 @@ public abstract class SyncPeerProtocolHandlerBase : ZeroProtocolHandlerBase, ISy
...
+        protected readonly MessageQueue<GetBlockBodiesMessage, (BlockBody[], long)> _bodiesRequests;
+
+        private readonly LatencyAndMessageSizeBasedRequestSizer _bodiesRequestSizer = new(
+            minRequestLimit: 1,
+            maxRequestLimit: 128,
+
+            // In addition to the byte limit, we also try to keep the latency of the get block bodies between these two
+            // watermark. This reduce timeout rate, and subsequently disconnection rate.
+            lowerLatencyWatermark: TimeSpan.FromMilliseconds(2000),
+            upperLatencyWatermark: TimeSpan.FromMilliseconds(3000),
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/P2P/ProtocolHandlers/SyncPeerProtocolHandlerBase.cs:66]: Maybe makes sense to make those values dependent on configured timeout? Or at least const?

File: src/Nethermind/Nethermind.Network/P2P/ProtocolHandlers/SyncPeerProtocolHandlerBase.cs [hunk 6]
```diff
@@ -51,7 +54,25 @@ public abstract class SyncPeerProtocolHandlerBase : ZeroProtocolHandlerBase, ISy
...
+
+            // When the bodies message size exceed this, we try to reduce the maximum number of block for this peer.
+            // This is for BeSU and Reth which does not seems to use the 2MB soft limit, causing them to send 20MB of bodies
+            // or receipts. This is not great as large message size are harder for DotNetty to pool byte buffer, causing
+            // higher memory usage. Reducing this even further does seems to help with memory, but may reduce throughput.
+            maxResponseSize: 3_000_000,
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/P2P/ProtocolHandlers/SyncPeerProtocolHandlerBase.cs:72]: reusable const?

File: src/Nethermind/Nethermind.Network/P2P/ProtocolHandlers/SyncPeerProtocolHandlerBase.cs [hunk 7]
```diff
@@ -51,7 +54,25 @@ public abstract class SyncPeerProtocolHandlerBase : ZeroProtocolHandlerBase, ISy
         protected readonly TxDecoder _txDecoder;
 
         protected readonly MessageQueue<GetBlockHeadersMessage, BlockHeader[]> _headersRequests;
-        protected readonly MessageQueue<GetBlockBodiesMessage, BlockBody[]> _bodiesRequests;
+        protected readonly MessageQueue<GetBlockBodiesMessage, (BlockBody[], long)> _bodiesRequests;
+
+        private readonly LatencyAndMessageSizeBasedRequestSizer _bodiesRequestSizer = new(
+            minRequestLimit: 1,
+            maxRequestLimit: 128,
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/P2P/ProtocolHandlers/SyncPeerProtocolHandlerBase.cs:61]: reusable const?

**Review by LukaszRozmej — APPROVED** [review]
Some minor improvements

File: src/Nethermind/Nethermind.Core/RequestSizer/LatencyAndMessageSizeBasedRequestSizer.cs (hunk 3 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Core/RequestSizer/LatencyAndMessageSizeBasedRequestSizer.cs:74]: Yes, the `_maxResponseSize`. But its already covered several line above. Also, lower watermark is lower than upper watermark, so if the latency is in between, (2 to 3 second), it does not change.

File: src/Nethermind/Nethermind.Network/P2P/ProtocolHandlers/SyncPeerProtocolHandlerBase.cs (hunk 6 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Network/P2P/ProtocolHandlers/SyncPeerProtocolHandlerBase.cs:72]: Yea, I did that first. But looks too verbose and less readable than just using the keyword args.

File: src/Nethermind/Nethermind.Network/P2P/ProtocolHandlers/SyncPeerProtocolHandlerBase.cs (hunk 5 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Network/P2P/ProtocolHandlers/SyncPeerProtocolHandlerBase.cs:66]: Maybe. Dont know. Does not have a very good idea on what to relate to.

File: src/Nethermind/Nethermind.Core/RequestSizer/LatencyAndMessageSizeBasedRequestSizer.cs (hunk 1 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Core/RequestSizer/LatencyAndMessageSizeBasedRequestSizer.cs]: done

File: src/Nethermind/Nethermind.Core/RequestSizer/LatencyAndMessageSizeBasedRequestSizer.cs (hunk 2 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Core/RequestSizer/LatencyAndMessageSizeBasedRequestSizer.cs]: done

**Pull request merged**